options.md
# Options
Options are like configs but they are set using a permissions plugin and support custom values for players and/or groups.
Options can also be defined in specific contexts (_server, world, claim, etc_) if supported by your permissions plugin.

## Available Options

| Option | Description | Default Value |
| ------ | ----------- | ------------- |
| `skyclaims.default-schematic` | The schematic that will be used when `is create` is used without a supplied schematic. | `skyfactory` |
| `skyclaims.default-biome` | The [Biome Type](biome-types.md) to use, if any, to set an island to on creation. | `none` |
| `skyclaims.min-size` | **Half** of the width of an island in blocks used to claim the player's usable space.<br /> _Accepts 8 - 256_ | `48` |
| `skyclaims.max-size` | **Half** of the max width of an island. Limits use of `/is expand` beyond this value.<br /> _Accepts **min-size** - 256_ | `64` |
| `skyclaims.max-spawns` | The maximum number of living entities allowed to spawn on islands owned by the player.<br /> _0 to disable_ | `70` |
| `skyclaims.max-spawns.hostile` | The maximum number of hostile entities allowed to spawn on islands owned by the player.<br /> _0 to disable_ | `50` |
| `skyclaims.max-spawns.passive` | The maximum number of passive entities allowed to spawn on islands owned by the player.<br /> _0 to disable_ | `30` |
| `skyclaims.expiration` | The number of days an island must be inactive before it is removed, if enabled.<br /> _0 to disable_ | `30` |
| `skyclaims.max-islands` | The maximum number of islands a player may join.<br /> _0 to disable_ | `0` |
| `skyclaims.max-teammates` | The maximum number of players that may join an island (_the island owner's limit is used_).<br /> _0 to disable_ | `0` |

## Examples
 
### LuckPerms
- `lp group default meta set skyclaims.default-schematic sf3` **&#8594;** islands will be made using sf3.schematic
- `lp group default meta set skyclaims.default-biome plains` **&#8594;** islands will be created as plains
- `lp group default meta set skyclaims.min-size 128` **&#8594;** Minimum island size will be 256x256 blocks
- `lp group default meta set skyclaims.max-size 240` **&#8594;** Maximum island size will be 480x480 blocks

CHANGELOG.md
# Change Log 

# Beta 28
### NOTE: Sponge Forge `1.12.2-2825-7.1.6-RC3697` adds entity and biome support for schematics
**REQUIRED: Sponge API 7.1 (SF 3682+); GP 1.12.2-4.3.0.622+; Permissions Plugin (ie. LuckPerms)**
**OPTIONAL: Nucleus 1.9.0-S7.1+**
- Added new schematic features:
  - `/is create` & `/is reset` now feature a chest UI when more than 1 schematic is available
  - `/is schematic` - schematic parent & list command
  - `/is schematic create <name>` - replaces `/isa createschematic <name>`
  - `/is schematic delete <schematic>` - deletes a schematic
  - `/is schematic command <schematic> <add|remove> <command>` - manages schematic commands
  - `/is schematic info <schematic>` - displays detailed information about a schematic
  - `/is schematic setbiome <schematic> <biome>` - set a default biome for a schematic that overrides the permission option
  - `/is schematic setheight <schematic> <height>` - set the generation height of a schematic
  - `/is schematic setname <schematic> <name>` - set an in-game name for a schematic that supports formatting code
  - `/is schematic seticon <schematic> <icon>` - set an icon for a schematic to be used by the chest GUI
  - `/is schematic setpreset <schematic> <preset>` - set a flat world preset for a schematic *_see flat world preset support_
  - `skyclaims.default-schematic` now defaults to empty which will list valid schematics
  - Removed `Misc-List-Schematics` config
  - Added `Misc > Text-Schematic-List` config to disable new chest UI
- Removed `/isa` &`/is admin`
  - `/is transfer` replaces `/isa transfer`
  - `/is reload` replaces `/isa reload`
  - `Admin-Command-Alias` config removed
- Added `skyclaims.max-teammates` option to limit the number of players per island
- Added fine-grained keep/clear inventory control:
  - Player inventory keep inventory permissions:
    - `skyclaims.keepinv.player.create`
    - `skyclaims.keepinv.player.delete`
    - `skyclaims.keepinv.player.kick`
    - `skyclaims.keepinv.player.leave`
    - `skyclaims.keepinv.player.reset`
  - EnderChest inventory keep inventory permissions:    
    - `skyclaims.keepinv.enderchest.create`
    - `skyclaims.keepinv.enderchest.delete`
    - `skyclaims.keepinv.enderchest.kick`
    - `skyclaims.keepinv.enderchest.leave`
    - `skyclaims.keepinv.enderchest.reset`
- Added `/is entity` command for detailed entity information
- Added `/is setname [name]` command
- Added `/scplayerinfo` command for debugging permission options
- Added support for flat world preset codes (_block ID portion only_) for region generation
  - See https://minecraft.gamepedia.com/Superflat#Preset_code_format for more details
- Added new schematics:
  - Stoneblock 2
  - SkyFactory 4
- Added `World > Regen-On-Create` config option
- Reworked `/is list [island] [sort type] [sort order]`
  - Sort order has been separated from sort type
  - **Sort Types**: NAME, CREATED, ONLINE, ACTIVE, MEMBERS, SIZE, ENTITIES
  - **Sort Orders**: ASC, DESC
  - Added `Misc >  Primary-List-Sort` config option - sets a sort type that gets applied before the one provided from the command argument
- Added `Misc > Island-Commands` config option
  - Commands trigger on island creation, join, and reset
  - Removed `Misc > Create-Commands` & `Misc > Reset-Commands`
- Removed outdated schematics:
  - Garden of Glass
  - SkyExchange
- Added island `min-size` check when an island owner logs in and if necessary, the island will be expanded
- Fixed schematics sometimes not generating at the intended height. The height set will be the height the player is at when standing on the lowest block of a schematic.
- Fixed Nucleus Integration commands not registering after a reload
- Fixed admin island expansion (`/is info`) bug where old clickable text can be used to expand outside the region
- Fixed a bug where `/is lock` would not prevent entry to a locked island
- Fixed a bug when removing overlapping claims during `/is create`
- Updated bStats to 1.4

# Beta 27.2 HOTFIX
**REQUIRED: SF build 2800+ & GP build 4.3.0.509+ OPTIONAL: Nucleus version 1.2.0+**
- **BREAKING CHANGES:**
    - Biome argument permissions have changed to `skyclaims.arguments.biomes.modid.biomeid`
    - Biome option changed to `modid.biomeid`
- **IMPORTANT:** fixed a bug where `/is info` could be used to delete an island you don't have permission to delete
- Fixed rare exception when duplicate modded biomes exist
- Added biome log feature to generate a list of supported biomes

# Beta 27.1
**REQUIRED: SF build 2800+ & GP build 4.3.0.509+ OPTIONAL: Nucleus version 1.2.0+**
- Fixed `/is home` not using safe teleport 
- Fixed `/is promote` & `/is demote` success messages 
- Fixed `/is expand` using more claim blocks than the owner has available
- Improved safe teleport failure messages
- Reworked the biome argument to _theoretically_ support modded biomes
- Removed `/isa config` 
- Removed Options Config - **Use permissions options instead**

# Beta 27.0
**REQUIRED: SF build 2800+ & GP build 4.3.0.509+ OPTIONAL: Nucleus version 1.2.0+**
- Subcommands now provide useful errors
- Updated for breaking API change in Sponge Forge 1.12.2-2555-7.0.0-BETA-2764
- Fixed `/is expand` erroring when an island argument was not provided
- Fixed `/is delete` requiring an island argument
- Fixed `/is delete` not removing players from the island before deletion
- Fixed wilderness flags applying to the default world when configured to use a different world
- Added `skyclaims.max-islands` option
- Added confirmation dialogs to `/is delete`
- Added configurable root command alias - the first alias is used for help text

# Beta 26.2 - S7.0 Hotfix
**REQUIRED: SF build 2688+ & GP build 4.3.0.473+ OPTIONAL: Nucleus version 1.2.0+**
- Fixed a bug with `/is lock` & `/is unlock` where the permission check returns incorrectly
- Fixed a typo in the `/is delete` help text
- Added a wilderness flag config setting to allows customizing the flags set automatically by SkyClaims

# Beta 26.1 - S7.0

src/main/java/net/mohron/skyclaims/permissions/Options.java
/*
 * SkyClaims - A Skyblock plugin made for Sponge
 * Copyright (C) 2017 Mohron
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SkyClaims is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with SkyClaims.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.mohron.skyclaims.permissions;

import java.util.Optional;
import java.util.UUID;
import net.mohron.skyclaims.SkyClaims;
import net.mohron.skyclaims.command.argument.BiomeArgument;
import net.mohron.skyclaims.schematic.IslandSchematic;
import org.apache.commons.lang3.StringUtils;
import org.spongepowered.api.service.permission.PermissionService;
import org.spongepowered.api.service.permission.Subject;
import org.spongepowered.api.world.biome.BiomeType;

public final class Options {

  private static final SkyClaims PLUGIN = SkyClaims.getInstance();
  private static final PermissionService PERMISSION_SERVICE = SkyClaims.getInstance().getPermissionService();

  // SkyClaims Options
  public static final String DEFAULT_SCHEMATIC = "skyclaims.default-schematic";
  public static final String DEFAULT_BIOME = "skyclaims.default-biome";
  public static final String MIN_SIZE = "skyclaims.min-size";
  public static final String MAX_SIZE = "skyclaims.max-size";
  public static final String MAX_SPAWNS = "skyclaims.max-spawns";
  public static final String MAX_HOSTILE = "skyclaims.max-spawns.hostile";
  public static final String MAX_PASSIVE = "skyclaims.max-spawns.passive";
  public static final String EXPIRATION = "skyclaims.expiration";
  public static final String MAX_ISLANDS = "skyclaims.max-islands";
  public static final String MAX_TEAMMATES = "skyclaims.max-teammates";

  private Options() {
  }

  public static Optional<IslandSchematic> getDefaultSchematic(UUID playerUniqueId) {
    String name = getStringOption(playerUniqueId, DEFAULT_SCHEMATIC, "");

    if(PLUGIN.getSchematicManager().getSchematics().size() == 1) {
      return Optional.of(PLUGIN.getSchematicManager().getSchematics().get(0));
    } else if (name.equalsIgnoreCase("random")) {
      return Optional.of(PLUGIN.getSchematicManager().getRandomSchematic());
    } else {
      return PLUGIN.getSchematicManager().getSchematics().stream()
          .filter(s -> s.getName().equalsIgnoreCase(name))
          .findAny();
    }
  }

  public static Optional<BiomeType> getDefaultBiome(UUID playerUniqueId) {
    String biomeOption = getStringOption(playerUniqueId, DEFAULT_BIOME, "");
    if (StringUtils.isBlank(biomeOption)) {
      return Optional.empty();
    }
    for (BiomeType biome : BiomeArgument.BIOMES) {
      if (biome.getId().equalsIgnoreCase(biomeOption) || biome.getId().equalsIgnoreCase("minecraft:" + biomeOption)) {
        return Optional.of(biome);
      }
    }
    return Optional.empty();
  }

  public static int getMinSize(UUID playerUniqueId) {
    return getIntOption(playerUniqueId, MIN_SIZE, 48, 8, 256);
  }

  public static int getMaxSize(UUID playerUniqueId) {
    return getIntOption(playerUniqueId, MAX_SIZE, 64, 8, 256);
  }

  public static int getMaxSpawns(UUID playerUniqueId) {
    return getIntOption(playerUniqueId, MAX_SPAWNS, PLUGIN.getConfig().getEntityConfig().getMaxSpawns());
  }

  public static int getMaxHostileSpawns(UUID playerUniqueId) {
    return getIntOption(playerUniqueId, MAX_HOSTILE, PLUGIN.getConfig().getEntityConfig().getMaxHostile());
  }

  public static int getMaxPassiveSpawns(UUID playerUniqueId) {
    return getIntOption(playerUniqueId, MAX_PASSIVE, PLUGIN.getConfig().getEntityConfig().getMaxPassive());
  }

  public static int getExpiration(UUID playerUniqueId) {
    return getIntOption(playerUniqueId, EXPIRATION, PLUGIN.getConfig().getExpirationConfig().getThreshold());
  }

  public static int getMaxIslands(UUID playerUniqueId) {
    return getIntOption(playerUniqueId, MAX_ISLANDS, 0);
  }

  public static int getMaxTeammates(UUID playerUniqueId) {
    return getIntOption(playerUniqueId, MAX_TEAMMATES, 0);
  }

  private static String getStringOption(UUID uuid, String option, String defaultValue) {
    Optional<Subject> subject = PERMISSION_SERVICE.getUserSubjects().getSubject(uuid.toString());
    return !subject.isPresent()
        ? defaultValue
        : subject.get().getOption(option).orElse(defaultValue);
  }

  private static int getIntOption(UUID uuid, String option, int defaultValue, int min, int max) {
    int value = getIntOption(uuid, option, defaultValue);
    return (value < min || value > max) ? defaultValue : value;
  }

  private static int getIntOption(UUID uuid, String option, int defaultValue) {
    Optional<Subject> subject = PERMISSION_SERVICE.getUserSubjects().getSubject(uuid.toString());
    if (!subject.isPresent()) {
      return defaultValue;
    }
    String value = subject.get().getOption(option).orElse(String.valueOf(defaultValue));
    try {
      return Integer.parseInt(value);
    } catch (NumberFormatException e) {
      return defaultValue;
    }
  }
}
